man/tpm2_tr_encode.1.md
% tpm2_tr_encode(1) tpm2-tools | General Commands Manual

# NAME

**tpm2_tr_encode**(1) - Encodes a peristent handle and `TPM2B_NAME` as a serialized `ESYS_TR` as
output.

# SYNOPSIS

**tpm2_tr_encode** [*OPTIONS*]

# DESCRIPTION

**tpm2_tr_encode**(1) - Encodes a peristent TPM2 handle along with a populated `TPM2B_PUBLIC` as
a serialized `ESYS_TR`. This is useful for moving a public and handle from one environment
where a TPM is not available to another environment with a TPM and make use of it through the
ESAPI API or tpm2-tools(1).

# OPTIONS

  * **-c**, **\--object-context**=_OBJECT_:

    Persistent handle.

[pubkey options](common/pubkey.md)

    Public key format.

  * **-o**, **\--output**=_FILE_:

    The output file path, recording the serialized `ESYS_TR`.


## References

[context object format](common/ctxobj.md) details the methods for specifying
_OBJECT_.

[common options](common/options.md) collection of common options that provide
information many users may expect.

[common tcti options](common/tcti.md) collection of options used to configure
the various known TCTI modules.

# EXAMPLES

## Serialize a public and handle as an ESYS_TR

```bash
tpm2_createprimary -c primary.ctx
tpm2_evictcontrol -c primary.ctx -o primary.tr 0x81000002
tpm2_readpublic -c primary.tr -o primary.pub
tpm2_tr_encode -c 0x81000002 -u primary.pub -o primary2.tr
```

[returns](common/returns.md)

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/misc/tpm2_tr_encode.c
/* SPDX-License-Identifier: BSD-3-Clause */

#include <stdbool.h>
#include <stdlib.h>
#include <stdint.h>

#include <tss2/tss2_mu.h>

#include "files.h"
#include "log.h"
#include "tpm2_convert.h"
#include "tpm2_identity_util.h"
#include "tpm2_tool.h"

typedef struct tpm2_tr_encode_ctx tpm2_tr_encode_ctx;
struct tpm2_tr_encode_ctx {
    const char *context_arg;
    const char *public_arg;
    const char *output_arg;
};
static tpm2_tr_encode_ctx ctx;

static bool tpm2_util_persistent_to_esys_tr(TPM2_HANDLE handle, TPM2B_PUBLIC *public, TPM2B_NAME *name, uint8_t **tr_buf, size_t *size) {

#define IESYSC_KEY_RSRC 1
    /*
     * So this is the format of an ESYS_TR:
     * 4 bytes TPM2_HANDLE
     * TPM2B_NAME
     * 4 bytes resource type
     * TPM2B_PUBLIC
     */

    /* can only serialize persistent objects */
    if ((handle >> TPM2_HR_SHIFT) != TPM2_HT_PERSISTENT) {
        LOG_ERR("Handle must be persistent, got: 0x%x", handle);
        return false;
    }

    /* Step 1 calculate the size */
    size_t buf_size = SIZE_MAX;
    size_t offset = 0;
    void *buffer = NULL;
    for (unsigned i=0; i < 2; i++) {
        TSS2_RC rc = Tss2_MU_TPM2_HANDLE_Marshal(
            handle,
            buffer,
            buf_size,
            &offset);
        if (rc != TSS2_RC_SUCCESS) {
            LOG_PERR(Tss2_MU_TPM2_HANDLE_Marshal, rc);
            free(buffer);
            return false;
        }

        rc = Tss2_MU_TPM2B_NAME_Marshal(
            name,
            buffer,
            buf_size,
            &offset);
        if (rc != TSS2_RC_SUCCESS) {
            LOG_PERR(Tss2_MU_TPM2B_NAME_Marshal, rc);
            free(buffer);
            return false;
        }

        rc = Tss2_MU_UINT32_Marshal(
            IESYSC_KEY_RSRC,
            buffer,
            buf_size,
            &offset);
        if (rc != TSS2_RC_SUCCESS) {
            LOG_PERR(Tss2_MU_UINT32_Marshal, rc);
            free(buffer);
            return false;
        }

        rc = Tss2_MU_TPM2B_PUBLIC_Marshal(
            public,
            buffer,
            buf_size,
            &offset);
        if (rc != TSS2_RC_SUCCESS) {
            LOG_PERR(Tss2_MU_TPM2B_PUBLIC_Marshal, rc);
            free(buffer);
            return false;
        }

        /*
         * on the first time through allocate the buffer for population on
         * the next loop
         */
        if (i == 0) {
            buf_size = offset;
            offset = 0;
            buffer = calloc(1, buf_size);
            if (!buffer) {
                return 1;
            }
        }
    }

    *size = offset;
    *tr_buf = buffer;

    return true;
}

static bool on_option(char key, char *value) {

    switch (key) {
    case 'c':
        ctx.context_arg = value;
        break;
    case 'u':
        ctx.public_arg = value;
        break;
    case 'o':
        ctx.output_arg = value;
        break;

    }
    return true;
}

static bool tpm2_tool_onstart(tpm2_options **opts) {
    const struct option topts[] = {
            { "object-context", required_argument, NULL, 'c' },
            { "public",         required_argument, NULL, 'u' },
            { "output",         required_argument, NULL, 'o' },
    };

    *opts = tpm2_options_new("c:u:o:", ARRAY_LEN(topts), topts, on_option,
          NULL, TPM2_OPTIONS_NO_SAPI);
    return *opts != NULL;
}

static bool check_options(void) {

    bool result = true;

    if (!ctx.context_arg) {
        LOG_ERR("Object Handle must be specified by option \"-c\"");
        result = false;
    }

    if (!ctx.public_arg) {
        LOG_ERR("Objects expected TPM2B_PUBLIC must be specified by option \"-u\"");
        result = false;
    }

    if (!ctx.output_arg) {
        LOG_ERR("The output file for the generated serialized ESYS_TR must be specified by option \"-o\"");
        result = false;
    }

    return result;
}

static tool_rc tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {

    UNUSED(flags);
    UNUSED(ectx);

    bool result = check_options();
    if (!result) {
        return tool_rc_option_error;
    }

    TPMI_RH_PROVISION handle;
    result = tpm2_util_handle_from_optarg(ctx.context_arg, &handle, TPM2_HANDLES_FLAGS_PERSISTENT);
    if (!result) {
        return tool_rc_option_error;
    }

    TPM2B_PUBLIC pub = { 0 };
    result = files_load_public(ctx.public_arg, &pub);
    if (!result) {
        LOG_ERR("Failed to load public key \"%s\"", ctx.public_arg);
        return tool_rc_option_error;
    }

    /* calculate the name */
    TPM2B_NAME name = TPM2B_TYPE_INIT(TPM2B_NAME, name);
    result = tpm2_identity_create_name(&pub, &name);
    if (!result) {
        LOG_ERR("Failed to calculate name");
        return tool_rc_general_error;
    }

    uint8_t *buf = NULL;
    size_t buf_size = 0;
    result = tpm2_util_persistent_to_esys_tr(handle, &pub, &name, &buf, &buf_size);
    if (!result) {
        LOG_ERR("Could not convert to serialized ESYS_TR");
        return tool_rc_general_error;
    }

    result = files_save_bytes_to_file(ctx.output_arg, buf,
            buf_size);
    free(buf);

    return result ? tool_rc_success : tool_rc_general_error;
}

// Register this tool with tpm2_tool.c
TPM2_TOOL_REGISTER("tr_encode", tpm2_tool_onstart, tpm2_tool_onrun, NULL, NULL)
